File: server/python/app/routes/flashcards.py
from flask import Blueprint, request, jsonify
from app.models.schemas import Flashcard
from app.utils.db_utils import get_flashcards, add_flashcard, delete_flashcard
from datetime import datetime

flashcards_bp = Blueprint("flashcards", __name__)


@flashcards_bp.route("/", methods=["GET"])
def get_all_flashcards():
    """Get all flashcards."""
    return jsonify(get_flashcards())


@flashcards_bp.route("/", methods=["POST"])
def create_flashcard():
    """Create a new flashcard."""
    data = request.json

    # Validate required fields
    required_fields = ["front", "back", "word"]
    for field in required_fields:
        if field not in data:
            return jsonify({"error": f"Missing required field: {field}"}), 400

    # Create flashcard with current timestamp
    flashcard = Flashcard(
        front=data["front"],
        back=data["back"],
        word=data["word"],
        exportedAt=datetime.now().isoformat(),
    )

    # Add to database
    result = add_flashcard(flashcard)

    return jsonify(result), 201


@flashcards_bp.route("/", methods=["DELETE"])
def remove_flashcard():
    """Delete a flashcard."""
    data = request.json

    # Validate required fields
    required_fields = ["front", "back", "word"]
    for field in required_fields:
        if field not in data:
            return jsonify({"error": f"Missing required field: {field}"}), 400

    # Delete from database
    success = delete_flashcard(data["word"], data["front"], data["back"])

    if success:
        return jsonify({"message": "Flashcard deleted successfully"}), 200
    else:
        return jsonify({"error": "Flashcard not found"}), 404


@flashcards_bp.route("/export", methods=["POST"])
def export_flashcards():
    """Export flashcards to various formats."""
    data = request.json

    if "flashcards" not in data or not isinstance(data["flashcards"], list):
        return jsonify({"error": "Invalid flashcards data"}), 400

    format_type = data.get("format", "anki")

    # Sanitize the flashcards data
    flashcards = []
    saved_flashcards = []
    for card in data["flashcards"]:
        flashcards.append(card)

        # Create a Flashcard object and save to database
        if "word" in card:
            word = card["word"]
            front = card["front"]
            back = card["back"]

            # Create flashcard with current timestamp if exportedAt is not provided
            export_time = card.get("exportedAt", datetime.now().isoformat())

            flashcard = Flashcard(
                front=front, back=back, word=word, exportedAt=export_time
            )

            # Add to database and collect results
            result = add_flashcard(flashcard)
            if result:
                saved_flashcards.append(result)

    if format_type == "anki":
        # Create Anki-compatible format
        # In a real app, this would generate an Anki package
        anki_data = {
            "notes": [
                {"front": card["front"], "back": card["back"], "tags": [card["word"]]}
                for card in flashcards
            ],
            "saved_flashcards": saved_flashcards,
        }
        return jsonify(anki_data)

    elif format_type == "csv":
        # Create CSV format (just the data structure, not the actual file)
        csv_data = {
            "headers": ["front", "back", "word"],
            "rows": [
                [card["front"], card["back"], card["word"]] for card in flashcards
            ],
            "saved_flashcards": saved_flashcards,
        }
        return jsonify(csv_data)

    else:
        return jsonify({"error": f"Unsupported format: {format_type}"}), 400

File: server/go/routes/flashcards.go
package routes

import (
	"log"
	"redefine/server/db"
	"redefine/server/types"
	"time"

	"github.com/gin-gonic/gin"
)

func setupFlashcardRoutes(api *gin.RouterGroup) {
	flashcardGroup := api.Group("/flashcards")

	flashcardGroup.GET("/", getAllFlashcards)

	flashcardGroup.POST("/", createFlashcard)

	flashcardGroup.DELETE("/", deleteFlashcard)

	flashcardGroup.POST("/export", exportFlashcards)
}

func getAllFlashcards(c *gin.Context) {
	flashcards, err := db.GetFlashcards()
	if err != nil {
		log.Printf("Error getting flashcards: %v", err)
		c.JSON(500, types.ErrorResponse{Error: "Failed to retrieve flashcards"})
		return
	}

	c.JSON(200, flashcards)
}

func createFlashcard(c *gin.Context) {
	var flashcard types.Flashcard
	if err := c.ShouldBindJSON(&flashcard); err != nil {
		c.JSON(400, types.ErrorResponse{Error: "Invalid flashcard data"})
		return
	}

	if flashcard.Front == "" || flashcard.Query == "" {
		c.JSON(400, types.ErrorResponse{Error: "Missing required fields"})
		return
	}

	if flashcard.ExportedAt == "" {
		flashcard.ExportedAt = time.Now().Format(time.RFC3339)
	}

	createdFlashcard, err := db.AddFlashcard(flashcard)
	if err != nil {
		log.Printf("Error adding flashcard: %v", err)
		c.JSON(500, types.ErrorResponse{Error: "Failed to create flashcard"})
		return
	}

	c.JSON(201, createdFlashcard)
}

func deleteFlashcard(c *gin.Context) {
	var request struct {
		Front string `json:"front"`
		Back  string `json:"back"`
		Query string `json:"query"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		c.JSON(400, types.ErrorResponse{Error: "Invalid request data"})
		return
	}

	if request.Front == "" || request.Query == "" {
		c.JSON(400, types.ErrorResponse{Error: "Missing required fields"})
		return
	}

	success, err := db.DeleteFlashcard(request.Query, request.Front, request.Back)
	if err != nil {
		log.Printf("Error deleting flashcard: %v", err)
		c.JSON(500, types.ErrorResponse{Error: "Failed to delete flashcard"})
		return
	}

	if !success {
		c.JSON(404, types.ErrorResponse{Error: "Flashcard not found"})
		return
	}

	c.JSON(200, gin.H{"message": "Flashcard deleted successfully"})
}

func exportFlashcards(c *gin.Context) {
	var request struct {
		Flashcards []types.Flashcard `json:"flashcards"`
		Format     string            `json:"format"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		log.Printf("Invalid request data: %v", err)
		c.JSON(400, types.ErrorResponse{Error: "Invalid request data"})
		return
	}

	if len(request.Flashcards) == 0 {
		c.JSON(400, types.ErrorResponse{Error: "No flashcards provided"})
		return
	}

	if request.Format == "" {
		request.Format = "anki"
	}

	log.Printf("Exporting %d flashcards in %s format", len(request.Flashcards), request.Format)

	var savedFlashcards []types.Flashcard
	var skippedCount int
	for _, card := range request.Flashcards {

		if card.Front == "" || card.Query == "" {
			log.Printf("Skipping flashcard with missing required fields: %+v", card)
			continue
		}

		if card.ExportedAt == "" {
			card.ExportedAt = time.Now().Format(time.RFC3339)
		}

		exists, err := db.FlashcardExists(card.Front, card.Back, card.Query)
		if err != nil {
			log.Printf("Error checking if flashcard exists: %v", err)
			continue
		}

		if exists {
			log.Printf("Flashcard already exists, skipping: %s / %s", card.Front, card.Query)
			skippedCount++

			savedFlashcards = append(savedFlashcards, card)
			continue
		}

		savedCard, err := db.AddFlashcard(card)
		if err != nil {
			log.Printf("Error saving flashcard: %v", err)

			continue
		}

		if savedCard != nil {
			log.Printf("Successfully saved flashcard: %s / %s", card.Front, card.Query)
			savedFlashcards = append(savedFlashcards, *savedCard)
		}
	}

	log.Printf("Export summary: %d saved, %d skipped (already existed)",
		len(savedFlashcards)-skippedCount, skippedCount)

	if request.Format == "anki" {

		ankiData := gin.H{
			"notes":            make([]gin.H, len(request.Flashcards)),
			"saved_flashcards": savedFlashcards,
		}

		for i, card := range request.Flashcards {
			ankiData["notes"].([]gin.H)[i] = gin.H{
				"front": card.Front,
				"back":  card.Back,
				"tags":  []string{card.Query},
			}
		}

		c.JSON(200, ankiData)
	} else if request.Format == "csv" {

		csvData := gin.H{
			"headers":          []string{"front", "back", "query"},
			"rows":             make([][]string, len(request.Flashcards)),
			"saved_flashcards": savedFlashcards,
		}

		for i, card := range request.Flashcards {
			csvData["rows"].([][]string)[i] = []string{card.Front, card.Back, card.Query}
		}

		c.JSON(200, csvData)
	} else {
		c.JSON(400, types.ErrorResponse{Error: "Unsupported format: " + request.Format})
	}
}
